[Feature/user-signup] 유저 기획 변경에 따른 API 재작업

Qbeom0925 opened the pull request:
## 작업 내용
유저 기획 변경에 따른 API 재작업이 진행되었습니다.
- 로그인 API
- 회원가입 API
- 이메일 전송 API
- 이메일 인증 확인 API
- 이미 가입된 유저인지 확인 API

## 관련 이슈
#58

Commit:
feature : 유저 회원가입 로그인 수정 중

Qbeom0925 committed:
refactor : 유저 API 재개발

- 로그인 API
- 회원가입 API
- 이메일 전송 API
- 이메일 인증 확인 API
- 이미 가입된 유저인지 확인 API

Qbeom0925 committed:
refactor : Spotless 적용

dldmsql commented on src/main/java/everymeal/server/user/controller/UserController.java:
```diff
@@ -36,86 +32,101 @@ public class UserController {
-    @Auth(require = true)
-    @SecurityRequirement(name = "bearerAuth")
-    @PostMapping("/email/auth/verify")
+    @PostMapping("/email/verify")
```
위의 이메일 전송 url의 규칙에 따라 `/verify/email` 로 명명하는 건 어떨까요?
명명규칙이 통일화되면 좋을 거 같아요.!

dldmsql commented on src/main/java/everymeal/server/user/controller/UserController.java:
```diff
@@ -36,86 +32,101 @@ public class UserController {
+        return ApplicationResponse.ok(userService.verifyEmailAuth(request));
+    }
+
+    @GetMapping("/email/check")
```
아니면 이메일 전송 부분을 `/email/send`로 변경한다거나..???

Qbeom0925 replied:
Restful하게 자원을 표현하는게 맞으니 
[POST] /email -> 전송
[GET] /email -> 중복확인
[GET] /email/verify -> 이메일 인증 확인 요 방식으로 가는 것은 어떨까요?!

dldmsql replied:
좋아요!!

dldmsql commented on src/main/java/everymeal/server/user/entity/User.java:
```diff
@@ -23,21 +24,29 @@ public class User extends BaseEntity {
 
-    private String deviceId;
-
+    @Column(unique = true)
     private String nickName;
 
+    @Column(unique = true)
     private String email;
```
nickName과 email 에 대해 조회가 빈번하게 이루어질 것 같으니, index를 거는 것에 대해 어떻게 생각하시나요???

Qbeom0925 replied:
` @Pattern(regexp = "^[^\\\\s]{1,10}$", message = "닉네임은 1~10자리로 입력해주세요.")
                @Schema(description = "이메일", example = "test@naver.com")
                String nickname`
회원가입 Requst에서 자리수를 규정하긴 하였으나 컬럼에 표현하는 것도 좋을 것 같습니다! 반영하겠습니다~

dldmsql commented on src/main/java/everymeal/server/user/entity/User.java:
```diff
@@ -23,21 +24,29 @@ public class User extends BaseEntity {
 
-    private String deviceId;
-
+    @Column(unique = true)
```
닉네임에 대한 데이터 크기를 지정하면 좋을 것 같아요. 

@Column(length = 50) 이런 식으로요!

Qbeom0925 replied:
` @Pattern(regexp = "^[^\\\\s]{1,10}$", message = "닉네임은 1~10자리로 입력해주세요.")
                @Schema(description = "이메일", example = "test@naver.com")
                String nickname`
회원가입 Requst에서 자리수를 규정하긴 하였으나 컬럼에 표현하는 것도 좋을 것 같습니다! 반영하겠습니다~

dldmsql commented on src/main/java/everymeal/server/user/entity/User.java:
```diff
@@ -23,21 +24,29 @@ public class User extends BaseEntity {
+    @Column(unique = true)
     private String nickName;
 
+    @Column(unique = true)
```
마찬가지로 이메일 계정에 대한 데이터 크기를 지정하면 좋을 것 같아요. `@Column(length = 100)`이런 식으로요!

데이터 크기 지정을 제안드리는 이유는 다음과 같아요.

1. 데이터 일관성을 위해 
2. 메모리 관점에서 공간 사용을 효율적으로 하기 위해

String으로 설정하면 기본 Varchar(255)로 지정되는데, 위의 데이터들 특성상 255 사이즈를 모두 채우는 일은 없을 거라 생각해서 공간을 최적화하기 위해 타이트하게 가져가면 좋을 것 같다는 생각입니다.

Qbeom0925 replied:
1번의 경우 일관성을 위해 좋은 생각인 듯 합니다!
하지만, 2번의 경우에서는 Varchar은 가변 자료형으로 저장된 문자열의 크기 만큼만을 저장하기에 Varchar 그 자체로 메모리에는 효율성을 가질 수 있습니다!~

Qbeom0925 replied:
참고해보시면 좋을 듯 합니다!
https://dev.mysql.com/doc/refman/8.0/en/char.html

dldmsql commented on src/main/java/everymeal/server/user/entity/User.java:
```diff
@@ -23,21 +24,29 @@ public class User extends BaseEntity {
+            University university) {
         this.nickName = nickName;
         this.email = email;
+        this.isDeleted = Boolean.TRUE;
```
유저에 대해 생성 시점에서 `isDeleated` 값을 TRUE로 하신 이유가 있나요..?? 반대..인 줄 알았어요

Qbeom0925 replied:
실수했네요..!!
ㅋㅋㅋㅋㅋ 감사합니다

dldmsql commented on src/main/java/everymeal/server/user/repository/UserRepository.java:
```diff
@@ -8,5 +8,7 @@
-    Optional<User> findByDeviceId(String deviceId);
+    Optional<User> findByEmail(String email);
+
+    Optional<Object> findByNickName(String nickname);
```
리턴 값으로 Object로 두신 이유가 있을까요..??

Qbeom0925 replied:
이것도 실수입니다...!

dldmsql commented on src/main/java/everymeal/server/user/service/UserService.java:
```diff
@@ -1,22 +1,21 @@
 public interface UserService {
 
-    Boolean signUp(UserSingReq userDeviceId);
+    UserLoginRes signUp(UserEmailSingReq userDeviceId);
```
param 변수명 통일화 어떠신가요..! `userDeviceId` -> `request`

Qbeom0925 replied:
api를 재개발하는 과정에서 이전에 사용한 메서드를 그대로 사용하다 보니 request 객체에 대해서는 변경하고 변수명에 대한 것을 신경 
못쎳네요 감사합니다.

dldmsql commented on src/main/java/everymeal/server/user/service/UserServiceImpl.java:
```diff
@@ -93,20 +124,16 @@ public UserEmailAuthRes emailAuth(
 
     @Override
     @Transactional
```
이메일 검증 로직에서 트랜잭션을 선언한 이유가 있을까요?? 디비와의 통신 로직이 없는 것 같아서요.!

Qbeom0925 replied:
해당 메서드도 이전 메서드를 재사용하면서 태랜잭션이 들어간 것 같군요! 감사합니다 !!

dldmsql commented on src/main/resources/application.yml:
```diff
@@ -10,7 +10,8 @@ spring:
   jpa:
     hibernate:
-      ddl-auto: none
+      ddl-auto: update
+
```
출시 이전까지는 update로 가는 건가용??

Qbeom0925 replied:
유저 엔티티 변경으로 인해서 update로 하였습니다!

dldmsql reviewed (changes requested):
작업 내용 확인했습니다!@ 수고하셨어요 👍🏻

Qbeom0925 committed:
refactor : Index 적용, 엔드포인트 재설정, 오타 수정

- User(email, nickname) 인덱스 적용
- repository Object-> User 오타 수정
- email 관련 엔드포인트 수정

Qbeom0925 committed:
refactor : Index 적용

- 닉네임 추가
